<a href="https://colab.research.google.com/github/AnnaBo1/python_for_DA/blob/main/hw_3_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [5]:
import pandas as pd
df = pd.read_csv('supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [6]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

In [7]:
profit = df.groupby('Branch')['gross income'].sum().reset_index() #без reset_index() видасть помилку. це метод для скидання індексу DataFrame
                                                                  #очислимо сумарний прибуток для кожного філіалу

most_profitable = profit.loc[profit['gross income'].idxmax()]     #.loc[], ми вибираємо рядок за індексом, який отримали раніше (idxmax()).
                                                                  #ми вибираємо рядок з найбільшим прибутком (найприбутковіший філіал).
                                                                  #branch_profit['gross income'].idxmax()
                                                                  #Ця частина коду знаходить індекс (номер рядка) з найбільшим значенням у стовпці 'gross income'. Метод idxmax() повертає індекс рядка, де дохід є максим
most_profitable
profit

Branch  gross income
0      A     5057.1605
1      B     5057.0320
2      C     5265.1765

Філіал С є найприбутковішим

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [8]:
profit = df.groupby(['Branch', 'City'])['gross income'].sum().reset_index()  #групуємо за філіалами і містами, і знаходимо сумарний прибуток по філіалах
most_profitable = profit.loc[profit['gross income'].idxmax()] #шуккаємо філіал з найвищим прибутком
most_profitable

Branch                  C
City            Naypyitaw
gross income    5265.1765
Name: 2, dtype: object

In [9]:
city_filter = df[df['City'] == 'Naypyitaw'] #фільтруємо DataFrame за конкретним містом

product = city_filter['Product line'].unique() #унікальні лінійки продуктів у цьому місті

product

array(['Electronic accessories', 'Home and lifestyle',
       'Food and beverages', 'Sports and travel', 'Health and beauty',
       'Fashion accessories'], dtype=object)

In [10]:
product_profit = df.groupby(['Product line','City'] )['gross income'].sum().reset_index() #groupby('Product line')['gross income'].sum():
                                                                                          #групуємо дані за лінійками продуктів і підраховуємо сумарний прибуток (gross income) для кожної лінійки.

product_profit_sorted = product_profit.sort_values(by='gross income', ascending=False) #sort_values(by='gross income', ascending=False):
                                                                                       #сортуємо результат у спадному порядку, щоб найбільш прибуткова лінійка була на першому місці.

most_profitable_product_line = product_profit_sorted.iloc[0] # iloc[0]: вибираємо перший рядок із найбільшим прибутком.Виводимо найбільш прибуткову лінійку продуктів

most_profitable_product_line

Product line    Food and beverages
City                     Naypyitaw
gross income              1131.755
Name: 7, dtype: object

Філіал з найвищим прибутком знаходиться у місті Naypyitaw.

Аналіз виявив, що найбільш прибутковою лінійко продуктів є Food and beverages. Тому є сенс відкрити ще один магазин Food and beverages в мітсі Naypyitaw.

Оскільки філіал C є найприбутковішим, варто звернути увагу на його  асортимент продукції та маркетинг.
Цей досвід може бути використаний для покращення результатів інших філіалів.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [11]:
invoices_by_category = df.pivot_table(index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')

invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [12]:
most_electronics = invoices_by_category['Electronic accessories'].idxmax()

most_electronics

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1) Як середній обсяг покупки змінюється в залежності від часу доби?

In [21]:
average_purchase_by_hour = df.groupby('Time')['Total'].mean() #Обчислюємо середнє значення для кожного часу в колонці Total
average_purchase_by_hour

Time
10:00    326.64100
10:01    102.97350
10:02    261.53750
10:03    335.01300
10:04    827.08500
           ...    
20:52    352.57950
20:54    187.80300
20:55    783.30000
20:57    148.99500
20:59    204.07275
Name: Total, Length: 506, dtype: float64

In [24]:
average_purchase_by_time = df.groupby('Time')['Total'].mean() #групуємо за колонкою 'Time' і обчислюємо середнє значення в колонці 'Total'
top_average_purchases = average_purchase_by_time.sort_values(ascending=False).head(5) #сортуємо середні значення по спаданню і вибираємо найбільші (наприклад, топ 5)
top_average_purchases

Time
16:20    1039.2900
16:18    1023.7500
14:50     931.0350
19:41     926.9505
14:45     914.5500
Name: Total, dtype: float64

Висновки: Час доби може впливати на середній розмір покупки.
Це допоможе вам швидко визначити, в які конкретні моменти часу обсяги покупок були найбільшими. Ця інформація може бути корисною для планування акцій або розподілу ресурсів у бізнесі і дозволяє визначити оптимальні години для розсилок і акцій.

2)Який вплив статі на популярність товарів?

In [46]:
top_5_products = df['Product line'].value_counts().head(5)
print("Топ-5 найпопулярніших товарів:")
top_5_products

Топ-5 найпопулярніших товарів:


Product line
Fashion accessories       178
Food and beverages        174
Electronic accessories    170
Sports and travel         166
Home and lifestyle        160
Name: count, dtype: int64

In [47]:
gender_counts = df.groupby(['Product line', 'Gender']).size()#підрахунок кількості покупок кожного товару по статі
print("Кількість покупок топ-5 товарів по статі:")
gender_counts

Кількість покупок топ-5 товарів по статі:


Product line            Gender
Electronic accessories  Female    84
                        Male      86
Fashion accessories     Female    96
                        Male      82
Food and beverages      Female    90
                        Male      84
Health and beauty       Female    64
                        Male      88
Home and lifestyle      Female    79
                        Male      81
Sports and travel       Female    88
                        Male      78
dtype: int64

Висновки: аналіз дозволяє легко порівнювати, скільки разів чоловіки і жінки купували кожен з топ-5 товарів. Інформація корисна для того, щоб адаптувати асортимент товарів, щоб задовольнити потреби обох статей. Адаптуйте асортимент товарів відповідно до переваг клієнтів. Наприклад, розширте асортимент у категорії 'Fashion accessories' для жінок і 'Electronic accessories' для чоловіків.

3)Чи є кореляція між кількістю покупок і середньою ціною товарів в кожній категорії?

In [50]:
correlation = df.groupby('Product line').agg({'Total': 'count', 'Unit price': 'mean', 'Quantity': 'sum'})
correlation                                #тотал Кількість записів у кожній категорії,Середня ціна товарів у кожній категорії,Сума кількості товарів у кожній категорії

Total  Unit price  Quantity
Product line                                       
Electronic accessories    170   53.551588       971
Fashion accessories       178   57.153652       902
Food and beverages        174   56.008851       952
Health and beauty         152   54.854474       854
Home and lifestyle        160   55.316937       911
Sports and travel         166   56.993253       920

In [51]:
correlation = df['Total'].corr(df['Unit price'])
print(f"Кореляція між кількістю покупок і середньою ціною товарів: {correlation:.2f}") #2f округлює число до двох десяткових знаків і відображає його у форматі з фіксованою комою

Кореляція між кількістю покупок і середньою ціною товарів: 0.63


Гіпотеза: Вища ціна товарів може знижувати частоту покупок у певних категоріях.

Висновок: Кореляція 0.63 показує помірний позитивний зв'язок між середньою ціною і кількістю покупок. Це свідчить про те, що в середньому, коли середня ціна товарів зростає, кількість покупок також має тенденцію зростати.
Оскільки існує позитивний зв'язок між середньою ціною і кількістю покупок, можливо, варто переглянути ціни. Наприклад, ви можете розглянути можливість підвищення цін на певні продукти, якщо це не призведе до значного зменшення кількості покупок. Це може допомогти збільшити дохід без значної втрати обсягу продажів.
Цей аналіз допомагає у прийнятті рішень маркетингових кампаній і управлінні асортиментом, що може в результаті підвищити дохід і задоволеність клієнтів.








